File: implement-shell-tools/ls/ls.py
import argparse
import os

parser = argparse.ArgumentParser(
    prog="check-for-ls",
    description="Implement my own version of ls",
)

parser.add_argument(
    "paths",
    nargs="*",
    help="The file paths to process"
)

parser.add_argument(
    "-1",
    "--one",
    action="store_true",
    help="List one file per line"
)

parser.add_argument(
    "-a",
    action="store_true",
    help="Show all files"
)

args = parser.parse_args()

paths = args.paths

if len(paths) == 0:
    paths = ["."]

for target in paths:

    if os.path.isdir(target):
        show_files = os.listdir(target)

        if args.a:
            show_files = [".", "..", *show_files]
        else:
            show_files = [
                name for name in show_files
                if not name.startswith(".")
            ]

        show_files.sort()

    else:
        show_files = [target]

    if args.one:
        for file in show_files:
            print(file)
    else:
        print("  ".join(show_files))

File: implement-shell-tools/ls/ls.js
const fs = require("fs");

const args = process.argv.slice(2);

let onePerLine = false;
let showHidden = false;
let paths = [];

for (let i = 0; i < args.length; i++) {
  if (args[i] === "-1") {
    onePerLine = true;
  } else if (args[i] === "-a") {
    showHidden = true;
  } else {
    paths.push(args[i]);
  }
}
if (paths.length === 0) {
  paths.push(".");
}

for (let i = 0; i < paths.length; i++) {
  let path = paths[i];

  try {
    if (fs.statSync(path).isFile()) {
      console.log(path);
    } else {
      let files = fs.readdirSync(path);

      files.sort();

      for (let j = 0; j < files.length; j++) {
        let file = files[j];
        if (!showHidden && file.startsWith(".")) {
          continue;
        }
        if (onePerLine) {
          console.log(file);
        } else {
          process.stdout.write(file + " ");
        }
      }

      if (!onePerLine) {
        process.stdout.write("\n");
      }
    }
  } catch (error) {
    console.log("Cannot access: " + path);
  }
}
